File: report-frontend/packages/report-api/src/datamodel.ts
import type { DataType } from './dataset';

// ============================================================
// Types（匹配后端 ReportConfigController 返回的 dataModel 结构）
// ============================================================

/** 数据源类型（对齐后端 DataSourceType 枚举） */
export type DataSourceType = 'CSV' | 'JSON' | 'DB' | 'API' | 'EXCEL';

/** JOIN 类型 */
export type JoinType = 'INNER' | 'LEFT' | 'RIGHT' | 'FULL';

export interface DataModelField {
  name: string;
  alias?: string;
  dataType: DataType;
  primaryKey?: boolean;
}

/** UNION 成员（对齐后端 UnionMemberDTO：统一列名 → 成员实际字段名） */
export interface UnionMember {
  datasetId: string;
  mapping?: Record<string, string>;
}

/**
 * 数据集（兼容两种端点返回，字段按需可选）：
 * - {@code GET /api/report/configs/{id}} 附带的 {@code dataModel.datasets}：精简视图
 *   {@code { id, alias, dataSourceType, fields }}（无 kind/datasourceId/sourceTable/members）
 * - {@code GET /api/datamodels/{id}} 返回的 {@code datasets}：完整 DatasetDTO
 *   {@code { id, alias, kind, datasourceId, sourceTable, fields, members }}（无 dataSourceType）
 */
export interface DataModelDataset {
  id: string;
  /** 英文标识名（TABLE=物理表名 sourceTable、UNION=合集名） */
  name?: string;
  alias?: string;
  /** 数据源类型（仅 configs/{id} 精简视图返回） */
  dataSourceType?: DataSourceType;
  /** 数据集形态（仅 datamodels/{id} 完整 DTO 返回）：TABLE | UNION */
  kind?: 'TABLE' | 'UNION';
  /** 物理表所属数据源 id（kind=TABLE） */
  datasourceId?: string;
  /** 物理表名/查询（kind=TABLE） */
  sourceTable?: string;
  /** UNION 成员（kind=UNION） */
  members?: UnionMember[];
  fields: DataModelField[];
}

/** 转换项条目（对齐后端 TransformItemDTO.EntryDTO：code/label/parent） */
export interface TransformEntryInfo {
  /** 原始编码（字段实际存储值） */
  code: string;
  /** 呈现文本（映射后展示） */
  label: string;
  /** 父级编码（树形用，顶层为空） */
  parent?: string;
}

/** 数据转换项（对齐后端 TransformItemDTO）：编码 → 呈现映射，报表中由 map(字段, 转换项id) 引用 */
export interface TransformItemInfo {
  id: string;
  /** 标识名（引用名） */
  name: string;
  /** 别名（中文名，展示用） */
  alias?: string;
  entries: TransformEntryInfo[];
}

export interface FieldRefInfo {
  datasetId: string;
  field: string;
}

export interface RelationshipInfo {
  left: FieldRefInfo;
  right: FieldRefInfo;
  joinType: JoinType;
}

/** 数据源连接（对齐后端 DataSourceDTO：{@code { id, name, type, config }}，敏感字段后端已脱敏） */
export interface DataModelSource {
  id: string;
  name: string;
  type: DataSourceType;
  /** 连接配置（含 path/url/凭证等，后端出口已脱敏） */
  config?: Record<string, unknown>;
}

/** 报表所用的数据模型：数据集 + 数据关系（报表参数属报表级，前端管理，不在此） */
export interface DataModelInfo {
  datasets: DataModelDataset[];
  relationships: RelationshipInfo[];
  /** 数据源连接列表（仅 GET /api/datamodels/{id} 完整返回；精简视图不含） */
  datasources?: DataModelSource[];
  /** 用户自定义转换项（在数据模型下配置，报表用 map() 引用） */
  transforms?: TransformItemInfo[];
}

File: report-engine-framework/src/main/java/com/codingapi/report/data/datamodel/DataModel.java
package com.codingapi.report.data.datamodel;

import com.codingapi.report.config.dto.ConfigDtos.FieldRefDTO;
import com.codingapi.report.config.dto.DataModelDtos.DataModelDTO;
import com.codingapi.report.config.dto.DataModelDtos.DataSourceDTO;
import com.codingapi.report.config.dto.DataModelDtos.DatasetDTO;
import com.codingapi.report.config.dto.DataModelDtos.FieldDTO;
import com.codingapi.report.config.dto.DataModelDtos.RelationshipDTO;
import com.codingapi.report.config.dto.DataModelDtos.UnionMemberDTO;
import com.codingapi.report.data.dataset.DataType;
import com.codingapi.report.data.dataset.Dataset;
import com.codingapi.report.data.dataset.Field;
import com.codingapi.report.data.dataset.FieldRef;
import com.codingapi.report.data.dataset.TableDataset;
import com.codingapi.report.data.dataset.UnionDataset;
import com.codingapi.report.data.dataset.UnionMember;
import com.codingapi.report.data.datasource.DataSource;
import com.codingapi.report.data.datasource.type.DataSourceType;
import com.codingapi.report.data.relation.JoinType;
import com.codingapi.report.data.relation.RelationOrigin;
import com.codingapi.report.data.relation.Relationship;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import lombok.Builder;
import lombok.Data;

/**
 * 数据模型（可复用的语义层）：描述"数据长什么样、表之间怎么关联"，<b>与具体报表无关</b>。
 *
 * <p>这是<b>领域对象</b>，也是 {@code DataModelRepository} 的持久化实体（不再有单独的 {@code DataModelConfig}）。 面向前端时通过
 * {@link #toDTO()} 转成 {@link DataModelDTO}（出口脱敏由上层做），前端 JSON 通过 {@link #fromDTO(DataModelDTO)} 还原。
 *
 * <h3>分层</h3>
 *
 * <p>建一次、多个 {@code Report} 引用（对齐 Power BI Tabular Model / Tableau Data Source / 帆软服务器数据集）。 数据计算全在 Java
 * 内存完成，连接（{@link DataSource}）只负责取规整表，故 {@link Relationship} 是内存 join 的 JoinSpec， 支持跨源关联（MySQL 表 JOIN
 * CSV 文件）。
 */
@Data
@Builder
public class DataModel {
    private String id;
    private String name;

    /** 状态（草稿/已发布）。 */
    private DataModelStatus status;

    /** 创建时间（epoch 毫秒）。 */
    private long createTime;

    /** 修改时间（epoch 毫秒）。 */
    private long updateTime;

    /**
     * 数据集列表：本模型选用的数据集，是报表绑定的最小粒度。
     *
     * <ul>
     *   <li>{@link TableDataset} —— 物理表，自带所属 {@link DataSource}（{@code dataset.getDatasource()}），
     *       故本模型不再单独持有 datasources 列表
     *   <li>{@link UnionDataset} —— UNION 派生（可跨源合并），定义在本模型内，无单一连接
     * </ul>
     */
    private List<Dataset> datasets;

    /** 跨数据集关联关系，所有引用本模型的报表共享（{@code RelationOrigin.AUTO} 自动扫描 / {@code MANUAL} 手动连线）。 */
    private List<Relationship> relationships;

    // ============================================================
    // 派生视图
    // ============================================================

    /** 去重收集本模型用到的连接（由各 TableDataset 自带的 datasource 聚合）。 */
    public List<DataSource> datasources() {
        Map<String, DataSource> byId = new LinkedHashMap<>();
        if (datasets != null) {
            for (Dataset ds : datasets) {
                if (ds instanceof TableDataset t && t.getDatasource() != null) {
                    byId.putIfAbsent(t.getDatasource().getId(), t.getDatasource());
                }
            }
        }
        return new ArrayList<>(byId.values());
    }

    // ============================================================
    // 领域 → DTO（出口；敏感字段脱敏由上层处理）
    // ============================================================

    public DataModelDTO toDTO() {
        return new DataModelDTO(
                id,
                name,
                status != null ? status.name() : null,
                createTime,
                updateTime,
                toDatasourceDtos(),
                toDatasetDtos(),
                toRelationshipDtos());
    }

    private List<DataSourceDTO> toDatasourceDtos() {
        List<DataSourceDTO> out = new ArrayList<>();
        for (DataSource s : datasources()) {
            out.add(
                    new DataSourceDTO(
                            s.getId(),
                            s.getName(),
                            s.getType() != null ? s.getType().type() : null,
                            s.getConfig()));
        }
        return out;
    }

    private List<DatasetDTO> toDatasetDtos() {
        List<DatasetDTO> out = new ArrayList<>();
        if (datasets == null) return out;
        for (Dataset ds : datasets) {
            if (ds instanceof TableDataset t) {
                out.add(
                        new DatasetDTO(
                                t.getId(),
                                t.getAlias(),
                                "TABLE",
                                t.getDatasourceId(),
                                t.getSourceTable(),
                                toFieldDtos(t.getFields()),
                                null));
            } else if (ds instanceof UnionDataset u) {
                out.add(
                        new DatasetDTO(
                                u.getId(),
                                u.getAlias(),
                                "UNION",
                                null,
                                null,
                                toFieldDtos(u.getFields()),
                                toUnionMemberDtos(u.getMembers())));
            }
        }
        return out;
    }

    private static List<FieldDTO> toFieldDtos(List<Field> fields) {
        List<FieldDTO> out = new ArrayList<>();
        if (fields == null) return out;
        for (Field f : fields) {
            out.add(
                    new FieldDTO(
                            f.getName(),
                            f.getAlias(),
                            f.getDataType() != null ? f.getDataType().name() : null,
                            f.isPrimaryKey()));
        }
        return out;
    }

    private static List<UnionMemberDTO> toUnionMemberDtos(List<UnionMember> members) {
        List<UnionMemberDTO> out = new ArrayList<>();
        if (members == null) return out;
        for (UnionMember m : members) {
            out.add(new UnionMemberDTO(m.datasetId(), m.mapping()));
        }
        return out;
    }

    private List<RelationshipDTO> toRelationshipDtos() {
        List<RelationshipDTO> out = new ArrayList<>();
        if (relationships == null) return out;
        for (Relationship r : relationships) {
            out.add(
                    new RelationshipDTO(
                            r.getId(),
                            toFieldRefDto(r.getLeft()),
                            toFieldRefDto(r.getRight()),
                            r.getJoinType() != null ? r.getJoinType().name() : null,
                            r.getOrigin() != null ? r.getOrigin().name() : null));
        }
        return out;
    }

    private static FieldRefDTO toFieldRefDto(FieldRef ref) {
        return ref == null ? null : new FieldRefDTO(ref.datasetId(), ref.field());
    }

    // ============================================================
    // DTO → 领域（入口）
    // ============================================================

    public static DataModel fromDTO(DataModelDTO dto) {
        if (dto == null) return null;
        Map<String, DataSource> sources = buildSources(dto.datasources());
        List<Dataset> datasets = buildDatasets(dto.datasets(), sources);
        attachDatasetsToSources(datasets, sources);
        return DataModel.builder()
                .id(dto.id())
                .name(dto.name())
                .status(dto.status() != null ? DataModelStatus.valueOf(dto.status()) : null)
                .createTime(dto.createTime())
                .updateTime(dto.updateTime())
                .datasets(datasets)
                .relationships(buildRelationships(dto.relationships()))
                .build();
    }

    private static Map<String, DataSource> buildSources(List<DataSourceDTO> sources) {
        Map<String, DataSource> out = new LinkedHashMap<>();
        if (sources == null) return out;
        for (DataSourceDTO s : sources) {
            DataSourceType type =
                    s.type() != null ? DataSourceType.of(s.type(), s.config()) : null;
            out.put(
                    s.id(),
                    DataSource.builder()
                            .id(s.id())
                            .name(s.name())
                            .type(type)
                            .config(s.config())
                            .build());
        }
        return out;
    }

    private static List<Dataset> buildDatasets(
            List<DatasetDTO> datasets, Map<String, DataSource> sources) {
        List<Dataset> out = new ArrayList<>();
        if (datasets == null) return out;
        for (DatasetDTO d : datasets) {
            if ("UNION".equals(d.kind())) {
                out.add(
                        UnionDataset.builder()
                                .id(d.id())
                                .alias(d.alias())
                                .fields(buildFields(d.fields()))
                                .members(buildUnionMembers(d.members()))
                                .build());
            } else {
                out.add(
                        TableDataset.builder()
                                .id(d.id())
                                .alias(d.alias())
                                .datasource(sources.get(d.datasourceId()))
                                .datasourceId(d.datasourceId())
                                .sourceTable(d.sourceTable())
                                .fields(buildFields(d.fields()))
                                .build());
            }
        }
        return out;
    }

    private static void attachDatasetsToSources(
            List<Dataset> datasets, Map<String, DataSource> sources) {
        Map<String, List<Dataset>> grouped = new LinkedHashMap<>();
        for (Dataset ds : datasets) {
            if (ds instanceof TableDataset t && t.getDatasource() != null) {
                grouped.computeIfAbsent(t.getDatasource().getId(), k -> new ArrayList<>()).add(t);
            }
        }
        grouped.forEach(
                (id, list) -> {
                    DataSource s = sources.get(id);
                    if (s != null) s.setDatasets(list);
                });
    }

    private static List<Field> buildFields(List<FieldDTO> fields) {
        List<Field> out = new ArrayList<>();
        if (fields == null) return out;
        for (FieldDTO f : fields) {
            out.add(
                    Field.builder()
                            .name(f.name())
                            .alias(f.alias())
                            .dataType(
                                    f.dataType() != null
                                            ? DataType.valueOf(f.dataType())
                                            : DataType.STRING)
                            .primaryKey(f.primaryKey())
                            .build());
        }
        return out;
    }

    private static List<UnionMember> buildUnionMembers(List<UnionMemberDTO> members) {
        List<UnionMember> out = new ArrayList<>();
        if (members == null) return out;
        for (UnionMemberDTO m : members) {
            out.add(new UnionMember(m.datasetId(), m.mapping() != null ? m.mapping() : Map.of()));
        }
        return out;
    }

    public static List<Relationship> buildRelationships(List<RelationshipDTO> rels) {
        List<Relationship> out = new ArrayList<>();
        if (rels == null) return out;
        for (RelationshipDTO r : rels) {
            out.add(
                    Relationship.builder()
                            .id(r.id())
                            .left(toFieldRef(r.left()))
                            .right(toFieldRef(r.right()))
                            .joinType(
                                    r.joinType() != null
                                            ? JoinType.valueOf(r.joinType())
                                            : JoinType.INNER)
                            .origin(
                                    r.origin() != null
                                            ? RelationOrigin.valueOf(r.origin())
                                            : RelationOrigin.MANUAL)
                            .build());
        }
        return out;
    }

    private static FieldRef toFieldRef(FieldRefDTO ref) {
        return ref == null ? null : new FieldRef(ref.datasetId(), ref.field());
    }
}
